### ISSUE TITLE
container_push fails with custom base image

### ISSUE THREAD

**Issue by marczych** [bazelbuild/rules_docker#450]
I had a working `container_push` target for a `py_image` but realized that I needed a custom base image because the code was shelling out to `git`. I got that all built, wired up, and working when running the image directly. However, my script to publish the image started failing with this error:

```
INFO: Running command line: bazel-bin/<redacted>
E0709 10:59:25.470009   19660 docker_session_.py:310] Error during upload of: <redacted>
F0709 10:59:25.470669   19660 fast_pusher_.py:171] Error publishing <redacted>: response: {'status': '400', 'content-length': '204', 'x-content-type-options': 'nosniff', 'server': 'nginx/1.9.15', 'connection': 'keep-alive', 'docker-distribution-api-version': 'registry/2.0', 'date': 'Mon, 09 Jul 2018 17:59:36 GMT', 'content-type': 'application/json; charset=utf-8'}
provided digest did not match uploaded content: {u'Reason': {}, u'Digest': u'sha256:717b092b8c86356cf976d9c83fa6f0ea67f2bf3148a5bbb7e02026a5d3245e68'}
ERROR: Non-zero return code '1' from command: Process exited with status 1
```

As far as I can tell, the only thing that changed from the perspective of the `container_push` target was the `base` argument to the `py_image` that it depends on.

I started poking around the push command and the runfiles to find that the computed digests indeed don't match:

```
[marc.zych@marc-desktop] - [~/.cache/bazel/_bazel_marc.zych/<redacted>/image] - [Mon Jul 09, 11:31:34]
[I]> sha256sum *.tar.gz
dfe544c43f03c0dbc7cc1dd0e0f0ff878010e5ca99dae949c246a8a5088c3b3d  000.tar.gz
296f72621ec779f9bec360cf247ef3baa5cdc3a374af19117b8d11cf6b5eebec  001.tar.gz
0de32c930a012885e189a14dc3fc6aed1853b23fca1d75810234f73ca2690b28  002.tar.gz
9a8c4414e2a0aa20e59beb6bd9cb92b1c56a25c52c5469b312741ee9ef11e0a9  003.tar.gz
481c50b91eb2bedac0866fced8f8588f8626d3e68d40db5a08c995edc7753472  004.tar.gz
f11797c11bb3af5d869e2a4408fa7d6fc6541c439c2d1cdba2dafde49ae4746b  005.tar.gz
[marc.zych@marc-desktop] - [~/.cache/bazel/_bazel_marc.zych/<redacted>/image] - [Mon Jul 09, 11:31:45]
[I]> for file in *.sha256; do
  echo $file: $(cat $file)
done
000.tar.sha256: 717b092b8c86356cf976d9c83fa6f0ea67f2bf3148a5bbb7e02026a5d3245e68
001.tar.sha256: 2df3095bdb63457521100a3ec9ae8bd9da8d75078a81d7df11b7988d3a98a0a1
002.tar.sha256: 9883b6d82ddfc1f002ff4b98c6198a7ed758ef0cf9631f2507e9c89ea544ae2c
003.tar.sha256: 176be293f240789763ad1304b5b629d80e0a6a6b69ab32c1192125c4b420fb71
004.tar.sha256: 83d4896a654c832dd6d8915ad0a1be3cc42c599bdda25029e85d29edaaf1cb59
005.tar.sha256: 5ff757f28895105288a17ac45b1c330dc2e5eb9d835c81698fc356d5ea68f392
```

I then looked at the runfiles for the working image to find some interesting `.tar.gz.nogz` files. But the digests _match_ here!

```
[marc.zych@marc-desktop] - [~/.cache/bazel/_bazel_marc.zych/<redacted>/image] - [Mon Jul 09, 11:33:29] - [1]
[I]> sha256sum *.nogz
3012758200fff06d102b078fe00161b9b152fcbcdae6e1ef182976f97057abde  000.tar.gz.nogz
2986b553e1e9268f4bb9ba3d226c95f2a4af505c6cf4e761a7f6ba37427e63fe  001.tar.gz.nogz
99a1cacbd3670590db736fec032470b81596c99ce58292e189df7deb36ab23f3  002.tar.gz.nogz
[marc.zych@marc-desktop] - [~/.cache/bazel/_bazel_marc.zych/<redacted>/image] - [Mon Jul 09, 11:33:43]
[I]> for file in *.sha256; do
  echo $file: $(cat $file)
done
000.tar.gz.nogz.sha256: 3012758200fff06d102b078fe00161b9b152fcbcdae6e1ef182976f97057abde
000.tar.gz.sha256: ca1df8c2ad9260ab07b7790e5ee9867dcc38ff2556ce3a4441a3d7f9b9bcc5d6
001.tar.gz.nogz.sha256: 2986b553e1e9268f4bb9ba3d226c95f2a4af505c6cf4e761a7f6ba37427e63fe
001.tar.gz.sha256: a79ed59a6f9f53ce631200429907b31aee38484f24589be8b65fe66a583a51c2
002.tar.gz.nogz.sha256: 99a1cacbd3670590db736fec032470b81596c99ce58292e189df7deb36ab23f3
002.tar.gz.sha256: 311d0d211592c05615ef31f9fdc7fd28a7ef6b2b8b3246477a7b081335125911
```

Some potentially relevant details:

 - Ubuntu 14.04
 - Bazel 0.7.3
 - rules_docker 473b7e9bdef8ef0a56c0fa3eead87c28ac6e007e
 - Custom base image was based on `python:2.7.15-alpine3.7` and added to bazel via `docker save --output ...` and `http_file`/`container_load` in `WORKSPACE`

**Comment by chrismcbride** [comment]
I have a similar problem. I believe its because the container_import rule generates the blobsum for the unzipped_layer:

https://github.com/bazelbuild/rules_docker/blob/0d6d69a2a4bbc33fc61a8350897b0e8136491ad5/container/import.bzl#L66

But the container_push rule uses the zipped layers:

https://github.com/bazelbuild/rules_docker/blob/0d6d69a2a4bbc33fc61a8350897b0e8136491ad5/container/push.bzl#L56

**Comment by nlopezgi** [comment]
Hi, sorry I have not had the time to look into this issue. If you could try to pursue more your hypothesis that this is related to how container_import generates the blobsum, and perhaps find a possible fix and send a PR it would be awesome. I'll try to look into it soon otherwise.

**Comment by nlopezgi** [comment]
Hi, is this now fixed with #478? Closing, but let me know otherwise to reopen.

**Comment by marczych** [comment]
Yes, this is fixed now. Thanks for following up!
